ruby/ql/lib/codeql/ruby/frameworks/data/internal/ApiGraphModelsSpecific.qll

```diff
@@ -66,9 +109,27 @@ API::Node getExtraNodeFromPath(string package, string type, AccessPath path, int
   )
 }
 
-/** Gets a Ruby-specific interpretation of the `(package, type)` tuple. */
-API::Node getExtraNodeFromType(string package, string type) {
-  isRelevantFullPath(package, type, _) and // Allow any package name, see `isPackageUsed`.
+/** Gets a Ruby-specific interpretation of the given `type`. */
+API::Node getExtraNodeFromType(string type) {
+  exists(string consts, string suffix, DataFlow::ConstRef constRef |
+    parseRelevantType(type, consts, suffix) and
+    constRef = getConstantFromConstPath(consts)
+  |
+    suffix = "!" and
```

Comment:
I wonder if we could use some different syntax, as `A::B!` would make me think _exatly_ type `A::B` (as opposed to `A::B` or any descendant thereof).

Reply:
Some other alternatives I've thought about
- `A::B.module`
- `A::B.class`
- `A::B.self`
- `A::B.instance` for instances and `A::B` for the class/module itself

For completeness there's also another alternative, which is to merge the `type` and `path` columns as well. So we'd have tables of form:
- `sourceModel(path, kind)`
- `sinkModel(path, kind)`
- `summaryModel(path, input, output, kind)`
- `typeModel(path1, path2)`

The "types" would then be written out as access paths, exactly as before
- `Member[A].Member[B]` for the `A::B` class/module itself
- `Member[A].Member[B].Instance` for instances of `A::B`

This means type model rows no longer just contribute to a named type, but instead contribute to an access path. I believe this is what @max-schaefer called an alias edge in some earlier discussions about this. If there isn't an intuitive syntax for the types you typically care about in Ruby, it might be worth trying to switch to that system instead.

Reply:
`A::B.static` maybe?

Reply:
If we want to use the same syntax across languages, then perhaps `A::B.static` is better.

Reply:
I think I perhaps favor the

> `A::B.instance` for instances and `A::B` for the class/module itself

suggestion, though that probably means having to write `.instance` a lot more often than having to write `.static` as per the suggestion above.

Reply:
Following along on the sidelines, two thoughts:

1) I thought it sounded like a fine idea to merge `type` and `path`, but how will we allow "fake" types used for aliasing, such as `~foo.Bar` (what we currently write as `foo;~Bar`)? (I know it's not going to happen right now, but just wanted to point out another potential problem)

2) I didn't personally have anything against using `!`. If we need to go with one of your alternatives, I would personally prefer `.class`. (I'm assuming that we'll do the same change in Python at some point, so that's why I dare participate in the coloring of this bikeshed :wink:) -- with my Python eyes:
    - `A::B.self` could easily be a reference to self parameters on methods of the `A::B` class
    - `A::B.module` could easily be a reference to the module containing `A::B`
    - I think `A::B.instance` is a natural idea, but wouldn't 99% of our type rows would be filled with `.instance`? I'm guessing it would, so don't think it's worth it.
    - `A::B.static` doesn't have any special meaning

Reply:
Or we we want to keep it closer to Ruby, `A::B.new` instead of `A::B.instance`.

Reply:
We could go with both `A::B.static` and `A::B.instance` to eliminate the ambiguity.

Reply:
> We could go with both `A::B.static` and `A::B.instance` to eliminate the ambiguity.

I like that compromise, for now.

Reply:
After implementing `.static` and `.instance` and seeing [what it looks like in practice](https://github.com/asgerf/codeql/commit/cb6d4c7a80e5ada0d0010851176e5bf7781df1fb) I don't really want to go ahead with that. I'd prefer to just stick with `!`.

> I thought it sounded like a fine idea to merge type and path, but how will we allow "fake" types used for aliasing, such as ~foo.Bar (what we currently write as foo;~Bar)? (I know it's not going to happen right now, but just wanted to point out another potential problem)

There would be a `Type[x]` token to support named types still, both for model-authored types, and type annotations whose meaning can't be known without seeing its definition (which is the case for TypeScript).